### PR TITLE
[4.0] Smart Search: Adding nested taxonomy

### PR DESCRIPTION
**Pull request by Hackwar**
Smart Search has a rather good taxonomy that allows to add a lot of additional filters for the search. However there are 2 flaws:

* Nested taxonomies (like categories) are squashed into a flat list and
* The state of a taxonomy is not updated in the search result.

This PR tries to fix both these flaws.

### How does it work?
For legacy finder plugins, everything stays as it was, so we should be pretty much backwards compatible here. However new (or updated) plugins can use the new ```FinderIndexerResult::addNestedTaxonomy()``` method to add nested taxonomies. This method gets a node object as a parameter, which has to implement at least the ImmutableNodeInterface interface, that is introduced with this PR. Internally it iterates down the tree along its parents, adds each parent as a taxonomy node with the previous node as its parent. In the result object, we are now also storing the taxonomy ID so that we can lookup the right taxonomy again later. The taxonomy table has been changed to a nested set table.

### How to test?
Apply these changes and make sure that the DB changes are applied, too. Then clear the index and re-run the indexing. You can then see the category filter in the frontend having a nested display, as well as in the backend. Check the taxonomy table in the DB for the nested entries, too.
Check if it properly filters for the respective taxonomies in the frontend when searching.
Enable displaying the taxonomies in the frontend and check that you can disable branches and single taxonomies and that those properly vanish from the filters and the result display. Create Search Filters in the backend and test those, too, if they properly honour the taxonomies.

### Status of this PR
This PR is mostly working, but I'm not 100% confident that I caught all changes that are necessary. I think I'm now depending a bit on feedback from you.

### CODE REVIEW

**Comment by Hackwar** [comment]
Right now this depends on #23452 being merged first.

File: administrator/components/com_finder/helpers/indexer/driver/mysql.php [hunk 1]
```diff
@@ -512,11 +487,11 @@ protected function toggleTables($memory)
 		elseif ($memory === false && $state !== false)
 		{
 			// Set the tokens table to MyISAM.
-			$db->setQuery('ALTER TABLE ' . $db->quoteName('#__finder_tokens') . ' ENGINE = MYISAM');
+			$db->setQuery('ALTER TABLE ' . $db->quoteName('#__finder_tokens') . ' ENGINE = INNODB');
```
Comment (HLeithner) [inline administrator/components/com_finder/helpers/indexer/driver/mysql.php:490]: Comment is wrong

File: administrator/components/com_finder/helpers/indexer/driver/mysql.php [hunk 2]
```diff
@@ -512,11 +487,11 @@ protected function toggleTables($memory)
...
 			$db->execute();
 
 			// Set the tokens aggregate table to MyISAM.
-			$db->setQuery('ALTER TABLE ' . $db->quoteName('#__finder_tokens_aggregate') . ' ENGINE = MYISAM');
+			$db->setQuery('ALTER TABLE ' . $db->quoteName('#__finder_tokens_aggregate') . ' ENGINE = INNODB');
```
Comment (HLeithner) [inline administrator/components/com_finder/helpers/indexer/driver/mysql.php:494]: Comment is wrong

File: administrator/components/com_finder/tmpl/maps/default.php [hunk 3]
```diff
@@ -79,19 +84,17 @@
 							</td>
 							<th scope="row">
 								<?php
-								if (trim($item->parent_title, '**') === 'Language')
+								/**if (trim($item->parent_title, '**') === 'Language')
```
Comment (HLeithner) [inline administrator/components/com_finder/tmpl/maps/default.php]: Is this needed in the future or whats the reason for commenting it out?

File: plugins/finder/contacts/contacts.php [hunk 4]
```diff
@@ -357,7 +357,9 @@ protected function index(FinderIndexerResult $item, $format = 'html')
 		$item->addTaxonomy('Type', 'Contact');
 
 		// Add the category taxonomy data.
-		$item->addTaxonomy('Category', $item->category, $item->cat_state, $item->cat_access);
+		$categories = jcategories::getInstance('com_contact');
```
Comment (HLeithner) [inline plugins/finder/contacts/contacts.php]:
```suggestion
		$categories = JCategories::getInstance('com_contact');
```

File: plugins/finder/content/content.php [hunk 5]
```diff
@@ -305,7 +305,9 @@ protected function index(FinderIndexerResult $item, $format = 'html')
 		}
 
 		// Add the category taxonomy data.
-		$item->addTaxonomy('Category', $item->category, $item->cat_state, $item->cat_access);
+		$categories = jcategories::getInstance('com_content');
```
Comment (HLeithner) [inline plugins/finder/content/content.php]:
```suggestion
		$categories = JCategories::getInstance('com_content');
```

File: plugins/finder/newsfeeds/newsfeeds.php [hunk 6]
```diff
@@ -293,7 +293,9 @@ protected function index(FinderIndexerResult $item, $format = 'html')
 		$item->addTaxonomy('Type', 'News Feed');
 
 		// Add the category taxonomy data.
-		$item->addTaxonomy('Category', $item->category, $item->cat_state, $item->cat_access);
+		$categories = jcategories::getInstance('com_newsfeeds');
```
Comment (HLeithner) [inline plugins/finder/newsfeeds/newsfeeds.php]:
```suggestion
		$categories = JCategories::getInstance('com_newsfeeds');
```

**Review by HLeithner — CHANGES_REQUESTED** [review]

**Comment by HLeithner** [comment]
On index or saving a article I get a fatal error:
```
Fatal error: Cannot use Joomla\CMS\Tree\NodeInterface as NodeInterface because the name is already in use in libraries/src/Categories/CategoryNode.php on line 18
```

File: administrator/components/com_finder/tmpl/maps/default.php (hunk 3 above)
Comment (Hackwar) [reply administrator/components/com_finder/tmpl/maps/default.php]: Sorry, fixed that one.

**Comment by Hackwar** [comment]
I've made all requested changes. Should be good to go. I can't see that the failing system test is related to this.

**Review by HLeithner — APPROVED** [review]

**Comment by wilsonge** [comment]
Looks like an improvement to me! Thanks!

**Comment by Hackwar** [comment]
:dancer: :champagne: Thank you!

**Pull request merged**